core-primitives/node-api/api-client-types/src/lib.rs

```diff
@@ -42,6 +44,39 @@ pub type ParentchainExtrinsicParamsBuilder = PlainTipExtrinsicParamsBuilder;
 pub type ParentchainUncheckedExtrinsic<Call> =
 	UncheckedExtrinsicV4<Call, SubstrateDefaultSignedExtra<PlainTip>>;
 
+/// Trait to extract signature and call indexes of an encoded [UncheckedExtrinsicV4].
+pub trait ExtractCallIndexAndSignature {
+	/// SignedExtra type of the Extrinsic.
+	type SignedExtra;
+
+	fn extract_call_index_and_signature(
+		encode_call: &mut &[u8],
+	) -> Option<(Signature<Self::SignedExtra>, CallIndex)>;
+}
+
+/// Signature type of the [UncheckedExtrinsicV4].
+pub type Signature<SignedExtra> = Option<(GenericAddress, MultiSignature, SignedExtra)>;
+
+impl<Call, SignedExtra> ExtractCallIndexAndSignature for UncheckedExtrinsicV4<Call, SignedExtra>
+where
+	// The Encode bounds are needed because of erroneous trait bounds in the api-client.
+	Call: Decode + Encode,
+	SignedExtra: Decode + Encode,
+{
+	type SignedExtra = SignedExtra;
+
+	/// Extract a call index of an encoded call.
+	///
+	/// Note: This mutates the slice. It will prune the `signature` field and the call
+	/// index of the `encoded_call`. Only the dispatchable's arguments are remaining.
+	fn extract_call_index_and_signature(
+		encoded_call: &mut &[u8],
+	) -> Option<(Signature<SignedExtra>, CallIndex)> {
+		let xt = UncheckedExtrinsicV4::<(CallIndex, ()), SignedExtra>::decode(encoded_call).ok()?;
+		Some((xt.signature, xt.function.0))
```

Comment:
I would do this instead.

Reply:
Thanks, I agree! I have updated the the interface to take an immutable borrow to the slice. However, I have made a local variable `let call_mut = &mut &call[...];` because we actually want that pointer to move to the beginning of the dispatchable's arguments to be decoded later.